pkg/datapath/loader/hash.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loader
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"hash"
+	"io"
+
+	"github.com/cilium/cilium/pkg/datapath"
+)
+
+var (
+	// DatapathSHA is set during build to the SHA generated from bindata.sh
+	// which is hashed across all datapath template code, excluding the
+	// node, netdev, lxc and sockops header files (see daemon/Makefile).
+	DatapathSHA string
+)
+
+// Hash represents a unique enumeration of the datapath implementation.
+type Hash struct {
+	hash.Hash
+}
+
+// NewHash creates a new datapath hash based on the contents of the datapath
+// template files under bpf/, generated by contrib/scripts/bindata.sh.
+func NewHash() *Hash {
+	h := sha1.New()
+	io.WriteString(h, DatapathSHA)
+	return &Hash{
+		Hash: h,
+	}
+}
+
+// HashDatapath returns a new datapath hash based on the specified datapath.
+//
+// The endpoint's static data is NOT included in this hash, for that perform:
+//	hash := HashDatapath(dp, nodeCfg, netdevCfg, ep)
+//	hashStr := hash.SumEndpoint(ep)
+func HashDatapath(dp datapath.Datapath, nodeCfg *datapath.LocalNodeConfiguration, netdevCfg datapath.DeviceConfiguration, epCfg datapath.EndpointConfiguration) *Hash {
+	h := NewHash()
+
+	// Writes won't fail; it's an in-memory hash.
+	if nodeCfg != nil {
+		_ = dp.WriteNodeConfig(h, nodeCfg)
+	}
+	if netdevCfg != nil {
+		_ = dp.WriteNetdevConfig(h, netdevCfg)
+	}
+	if epCfg != nil {
+		_ = dp.WriteTemplateConfig(h, epCfg)
+	}
+
+	return h
+}
+
+// Copy the cached mash of the hashed datapath paths.
+func (h *Hash) Copy() *Hash {
+	newHash := sha1.New()
+	newHash.Write(h.Data())
```

Comment:
This does not create a Copy, this is creating a new Hash with a different sum!

Reply:
Sigh, true. The base hash libraries didn't seem to do what I wanted them to do so this is basically an attempt to hack a copy operation in. Really it's a `New()` operation that derives from the same hash, which was deterministic enough at least to get the basic template hashing to work consistently -- ie the sum is the same between multiple endpoints on which you call `baseHash.SumEndpoint(...)` with the same input.

Do you have any suggestions for a saveable hash structure library where you don't have to do either of the following?
* Share the same hash object (lock and contend with other goroutines)
* Rehash the same base data again and again

Ideally we could also seed it given that we have already performed the first portion of the hashing all the way back in the build process with bpf.sha.

Reply:
OK, I was missing that with `encoding.BinaryMarshaler`, standard library hashes can be serialized then unserialized to a new value and the new value returned, which is equivalent to making a copy.